src/lib.rs

```diff
@@ -101,9 +116,58 @@ static MIN_TIME_PERIOD: once_cell::sync::Lazy<winapi::shared::minwindef::UINT> =
 pub(crate) fn thread_sleep(duration: Duration) {
     unsafe {
         use winapi::um::timeapi::{timeBeginPeriod, timeEndPeriod};
+
+        /*
+        This is a *BIG* no no..  The problem here is that timeBeginPeriod has
+        global impact on the entire OS.  The overhead of changing this value
+        regularly is ... not good, since all processes (not just threads) in the
+        system will now have different accuracy for sleeps, system interrupt
+        rates are increased, etc etc.  Not to mention that the cost of the pair
+        of calls is exceptionally high.  From the docs:
+
+        "Use caution when calling timeBeginPeriod, as frequent calls can
+        significantly affect the system clock, system power usage, and the scheduler."
+
+        To keep using this, only do it once a frame at most and even that is
+        not really a great idea since it can trigger a number of bugs in the
+        overall system where timers change accuracy unexpectedly in completely
+        unrelated processes.
+         */
         timeBeginPeriod(*MIN_TIME_PERIOD);
         thread::sleep(duration);
         timeEndPeriod(*MIN_TIME_PERIOD);
+
+        /*
+        The option to replace this is a spin wait using _mm_pause.  This is a
+        low level instruction you can find in:
+        https://doc.rust-lang.org/1.29.1/core/arch/x86_64/fn._mm_pause.html
+
+        There are two problems with this item though, one building on the other.
+        Intel, in their infinite wisdom, decided to change the behavior of the
+        silicon on Skylake and broke a lot of things thanks to the change.
+        Everything prior to Skylake used a delay of between 6 and 14 CPI,
+        Skylake cranked the hell out of it and it now delays the CPU by 140 CPI.
+
+        The second problem follows from the change: for tight multicore work that
+        targets multiple CPU's now have to detect if the CPU is Skylake to
+        adjust spin wait repetitions or risk massive wait time behavior changes.
+
+        Anyway, you would end up with something like:
+        loop {
+            let count = 32; // Pre-Skylake
+            for i in 0..count {
+                std::arch::x86_64::_mm_pause;
+            }
+
+            // Test current time.
+            // As a note, calling time functions is also slow, so
+            // generally you want a linear backoff where the count
+            // above is about 1/2 the "guessed" at length of the
+            // desired sleep period the first time through, then
+            // use half of that again for the next loop etc till
+            // you get "close enough".
+        }
```

Comment:
Interesting idea, but on the face of it it seems a bit of a pain. I'd rather avoid hardware specific stuff if I can.

Reply:
Well, unfortunately in a number of areas of games, being a little hardware centric is pretty much a requirement since OS's are not really intended to make games.  If they were, we wouldn't need this sort of "fix" for something so simple now would we?  :)  Anyway, this is basically an outline of what I've had to do on every game I've shipped over the last 20+ years because otherwise you incur too much OS overhead, have very inaccurate frame rate limiting or generally end up just slowing everything down in general.  The problem with Intel changing the delay for pause is unfortunately a massive hassle and my friends at Intel actually mentioned that it was a bug they didn't catch till it was too late, so they tried to play it as "we intended that"....  Assholes...